### PR TITLE
docs: Rolling upgrades and cluster config in Kubernetes

### LINKED ISSUE #950: Provide a note on any impacts on clients / or any issues might be seen during Redpanda "upgrades"
### Describe the Issue
This is to cover all upgrades (so might be separate places in documentation, and can spin out new issues if needed for dif doc sections  )

- helm chart upgrade 
- Cloud "self-service" upgrades
- Upgrades performed by Redpanda on cloud clusters 
...any other mechanism that upgrades the current version of Redpanda to a higher one.

It is a very common (and reasonable ) question that is asked.... 
what is the impact of the upgrade especially for clients when redpanda brokers are restarted in clusters

Lets document this, so admins/customers are fully aware of what to expect ...
rather than having to log support tickets to get the answer..
 
Impact to clients. (producers, consumers + schema registry client) 
errors etc that might be seen as/after RP restarted... and how some it will depend on the retry logic of clients..
(realise more detail needed for docs team here.. but want this logged )

Also document things like 
- might see an initial spike in CPU seen after a broker comes back.. as it deals with some rebalancing/leadership stuff etc
- possible spikes in errors in logs etc (rebalancing/leadership stuff) which might trigger customers SRE alerting systems

### Updates to existing documentation
This is to cover all upgrades (so might be separate places in documentation, and can spin out new issues if  )
### If new feature, in which release is it included or expected? 
not a new feature in "general" but self-service upgrades in cloud is ..
 
### Additional notes
Internal  Incident 118 . Dev confirmed initial spike in CPU
Will add additional stuff on connection errors/retry stuff..

### LINKED ISSUE #1170: Add rolling restart guide for Kubernetes deployments to docs
### Describe the Issue

I Redpanda user requested details on how to do rolling restarts for a Redpanda cluster within Kubernetes, and I didn't see any documentation on this. We have the following two docs:
https://docs.redpanda.com/docs/manage/cluster-maintenance/rolling-upgrade/
https://docs.redpanda.com/docs/manage/node-management/

Each covers a portion of the process for rolling restarts, but neither gives details specific to Kubernetes deployments.

### Updates to existing documentation

It's possible either of the links above could be modified to include the details I've included in this issue.

### New feature or content gap requests

It's also possible this is a content gap that deserves it's own new page. We have a Redpanda University course that covers this content fairly well, but again it does not mention Kubernetes.

### Additional notes

Below is what I have prepared for the Redpanda user who requested this info, and the documentation change could incorporate this content:

===

### Externally connect via local rpk
In order to use `rpk` from your local computer, you need to know how to access the Kubernetes cluster externally. Find your external advertised broker addresses and ports by first running the following command:
```
kubectl exec -it -n redpanda redpanda-0 -c redpanda -- rpk cluster info --brokers localhost:9092
```

Output:
```
CLUSTER
=======
redpanda.c2b00813-0c97-4622-a377-236f593ec915

BROKERS
=======
ID    HOST        PORT
0*    redpanda-0  31092
1     redpanda-1  31092
2     redpanda-2  31092
```

In the above output we see that each broker advertises port 31092 (which is the default port when using the NodePort service).

> Note: Replace `9092` with whatever you have defined for the external kafka API traffic. If you have deployed Redpanda via the helm chart, it defaults to `9094`.

> Note: Make sure that you aren't mistakenly sending the above command to the port for *internal* kafka API traffic! If you have deployed Redpanda via the helm chart, it defaults to `9093`.

It is helpful to set `REDPANDA_BROKERS` for rpk commands that use the kafka API:
``` 
export REDPANDA_BROKERS=redpanda-0:31092,redpanda-1:31092,redpanda-2:31092
```

You should now be able to use your local `rpk` to access your cluster. If you run into issues, see this [link](https://docs.redpanda.com/docs/manage/kubernetes/networking/networking-and-connectivity/) for more details and help.

### Following logs
[stern](https://github.com/stern/stern) is a logging tool that makes tailing Kubernetes resources easy. Once installed, run the following command to tail all Redpanda brokers:
```
stern -A redpanda-
```

### Artificially requiring a restart
These instructions from this point assume your cluster is in a healthy state without any brokers needing a restart. If you already know your brokers require a restart, then skip to the "Rolling restart" section below.

Check the cluster status to verify no brokers require restart:
```
rpk cluster config status --api-urls redpanda-0:31644,redpanda-1:31644,redpanda-2:31644
```

Output:
```
NODE  CONFIG-VERSION  NEEDS-RESTART  INVALID  UNKNOWN
0     3               false          []       []
1     3               false          []       []
2     3               false          []       []
```

In order artificially require a restart, we will make a change to a cluster-wide property that requires a restart in order to take affect. Enable `aggregate_metrics` on your cluster by editing this property's value with the following command:
```
rpk cluster config edit --api-urls redpanda-0:31644,redpanda-1:31644,redpanda-2:31644
```

Search for `aggregate_metrics` and set it's value to `true`. Once done, save the file and exit the editor.

> Note: Details on `aggregate_metrics` can be found [here](https://docs.redpanda.com/docs/next/reference/cluster-properties/#aggregate_metrics)

Check the cluster status to see that all brokers now require restart:
```
rpk cluster config status --api-urls redpanda-0:31644,redpanda-1:31644,redpanda-2:31644
```

Output:
```
NODE  CONFIG-VERSION  NEEDS-RESTART  INVALID  UNKNOWN
0     3               true           []       []
1     3               true           []       []
2     3               true           []       []
```

## Rolling restart
Check which broker is the controller:
```
rpk cluster info
```

Output:
```
CLUSTER
=======
redpanda.c2b00813-0c97-4622-a377-236f593ec915

BROKERS
=======
ID    HOST        PORT
0*    redpanda-0  31092
1     redpanda-1  31092
2     redpanda-2  31092
```

In this example, ID 0 is the controller. Restarting the controller last ensures control changes to another broker only once during a rolling restart. Put the broker with node-id 1 in maintenance mode:
```
rpk cluster maintenance enable 1 --wait --api-urls redpanda-0:9644,redpanda-1:9644,redpanda-2:9644
```

Ensure broker 1 is in maintenance mode (`FINISHED: true`):
```
rpk cluster maintenance status --api-urls redpanda-0:9644,redpanda-1:9644,redpanda-2:9644
```

Once broker 1 is in maintenance mode, you can then restart. One way to do this is with the following command:
```
kubectl exec -it -n redpanda redpanda-1 -c redpanda -- rpk redpanda stop
```

Check the status to make sure broker 1 no longer requires restart:
```
rpk cluster config status --api-urls redpanda-0:31644,redpanda-1:31644,redpanda-2:31644
```

Take broker 1 out of maintenance mode:
```
rpk cluster maintenance disable 0 --api-urls redpanda-0:31644,redpanda-1:31644,redpanda-2:31644
```

Wait until broker 1 is out of maintenance mode (`FINISHED: false`):
```
rpk cluster maintenance status --api-urls redpanda-0:9644,redpanda-1:9644,redpanda-2:9644
```

The process for one broker is complete. Repeat for the other brokers, and it's recommended to restart the controller broker last.

### LINKED ISSUE #1296: Rolling upgrades in Kubernetes
### Describe the Issue

We need a dedicated guide for doing rolling upgrades in Kubernetes.

### PR DESCRIPTION
**Pull request by JakeSCahill**
Review deadline: March 2

Fixes #1296, fixes #1170, fixes #950

Previews:

- [Cluster Properties](https://deploy-preview-1297--redpanda-documentation.netlify.app/docs/beta/manage/kubernetes/cluster-property-configuration/)
- [Rolling Upgrades](https://deploy-preview-1297--redpanda-documentation.netlify.app/docs/beta/manage/kubernetes/rolling-upgrade/)

### CODE REVIEW

File: docs/manage/kubernetes/rolling-upgrade.mdx [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+
+## Rollbacks
+
+If something does not go as planned during a rolling upgrade, you can roll back to the original version as long as you have not upgraded every broker. Replace `<original-version>` with your original version.
+
+```
+helm upgrade --install redpanda redpanda/redpanda \
+  --namespace redpanda \
+  --create-namespace \
+  --set image.tag=<original-version>
```
Comment (joejulian) [inline docs/manage/kubernetes/rolling-upgrade.mdx:415]: We should test using the `helm rollback` command. It should do this correctly.

**Review by joejulian — APPROVED** [review]
Approving as there's nothing wrong with the commands, but I do think it would be a good idea to look at and consider using `helm rollback`. It reduces the chance of user-error making things worse.

File: docs/manage/kubernetes/rolling-upgrade.mdx (hunk 1 above)
Comment (JakeSCahill) [reply docs/manage/kubernetes/rolling-upgrade.mdx:415]: confirmed that `helm rollback` works. @joejulian @alejandroEsc @RafalKorepta are we all happy to make this the recommended procedure?

Comment (JakeSCahill) [reply docs/manage/kubernetes/rolling-upgrade.mdx]: confirmed on Slack

File: docs/manage/cluster-maintenance/cluster-property-configuration.mdx [hunk 2]
```diff
@@ -1,20 +1,34 @@
...
     <meta name="description" content="Configure cluster properties using rpk commands" data-rh="true"/>
 </head>
 
+import ContextLink from '@site/src/components/ContextButton.js'
+
+<ContextLink frontmatter={frontMatter}
+/>
+
 When you install Redpanda software, it automatically generates a cluster configuration with default settings assigned to each cluster property. 
 You might want to change some of the default settings, particularly if you plan to use features such as rack awareness or tiered storage, which are disabled by default.
```
Comment (micheleRP) [inline docs/manage/cluster-maintenance/cluster-property-configuration.mdx:23]:
```suggestion
You might want to change some of the default settings, particularly if you plan to use features such as rack awareness or Tiered Storage, which are disabled by default.
```

File: docs/manage/kubernetes/cluster-property-configuration.mdx [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+import ClusterProperty from '../shared/tiered-storage/_cluster-props.mdx'
+import ContextLink from '@site/src/components/ContextButton.js'
+
+<ContextLink frontmatter={frontMatter}
+/>
+
+When you install Redpanda, it automatically generates a cluster configuration with default settings assigned to each cluster property. 
+You might want to change some of the default settings, particularly if you plan to use features such as rack awareness or Tiered Storage, which are disabled by default.
+
+To configure cluster properties, you must override the defaults in the Redpanda Helm chart's values, using the `helm upgrade --install` command.
```
Comment (micheleRP) [inline docs/manage/kubernetes/cluster-property-configuration.mdx:29]:
```suggestion
To configure cluster properties, you must override the defaults in the Redpanda Helm chart's values using the `helm upgrade --install` command. 
```

File: docs/manage/kubernetes/cluster-property-configuration.mdx [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+To view all cluster properties, along with their descriptions and default settings, refer to [Cluster configuration properties](../../../reference/cluster-properties) and [Tunable properties](../../../reference/tunable-properties). Some cluster configuration properties are nested within other objects in the Helm chart's values. For example, `cloud_storage_access_key` is nested inside `storage.tieredStorage`.
+
+:::note
+[Tunable properties](../../../reference/tunable-properties), such as internal buffer sizes, have default settings that you typically do not need to modify during normal operations. You should only change these properties after collecting sufficient data to warrant a change. 
+:::
+
+## Edit cluster properties
+
+To change any property settings, edit the default configuration using the `--values` or `--set` flags in the `helm upgrade --install` commmand. For example, to set the <ClusterProperty name="cloud_storage_access_key" frontmatter={frontMatter}/> cluster configuration:
```
Comment (micheleRP) [inline docs/manage/kubernetes/cluster-property-configuration.mdx:42]:
```suggestion
To change any property settings, edit the default configuration using the `--values` or `--set` flags in the `helm upgrade --install` command. For example, to set the <ClusterProperty name="cloud_storage_access_key" frontmatter={frontMatter}/> cluster configuration:
```

File: docs/manage/kubernetes/cluster-property-configuration.mdx [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+
+```
+producer
+```
+
+</details>
+
+## Export a Redpanda configuration file
+
+If you want to see all Redpanda configuration for a broker, you can use the `rpk cluster config export` command to save the current Redpanda configuration to a file. For example, you may want to use the configuration file during debugging.
```
Comment (micheleRP) [inline docs/manage/kubernetes/cluster-property-configuration.mdx:115]:
```suggestion
If you want to see all Redpanda configurations for a broker, you can use the `rpk cluster config export` command to save the current Redpanda configuration to a file. For example, you may want to use the configuration file during debugging.
```

Comment (micheleRP) [inline docs/manage/kubernetes/cluster-property-configuration.mdx:115]:
```suggestion
To see all Redpanda configuration for a broker, you can use the `rpk cluster config export` command to save the current Redpanda configuration to a file. For example, you may want to use the configuration file during debugging.
```

File: docs/manage/cluster-maintenance/cluster-property-configuration.mdx [hunk 6]
```diff
@@ -1,20 +1,34 @@
...
-[Tunable properties](../../../reference/tunable-properties), such as internal buffer sizes, have default settings that you typically do not need to modify during normal operations. You should only change these properties after collecting sufficient data to warrant a change. 
-
 To view all cluster properties, along with their descriptions and default settings, refer to [Cluster configuration properties](../../../reference/cluster-properties) and [Tunable properties](../../../reference/tunable-properties).
 
-## Editing cluster properties
+:::note
+[Tunable properties](../../../reference/tunable-properties), such as internal buffer sizes, have default settings that you typically do not need to modify during normal operations. You should only change these properties after collecting sufficient data to warrant a change. 
+:::
+
+## Edit cluster properties
```
Comment (micheleRP) [inline docs/manage/cluster-maintenance/cluster-property-configuration.mdx:31]: @JakeSCahill, while you're in this file, can you please align the content in steps 5 & 6 here?

File: docs/manage/kubernetes/troubleshooting/troubleshoot.mdx [hunk 7]
```diff
@@ -144,7 +144,7 @@ This error appears when a cluster has TLS enabled, and you try to access the adm
 Retrying POST for error: Post "http://127.0.0.1:9644/v1/security/users": net/http: HTTP/1.x transport connection broken: malformed HTTP response "\x15\x03\x03\x00\x02\x02"
```
Comment (micheleRP) [inline docs/manage/kubernetes/troubleshooting/troubleshoot.mdx:144]: @JakeSCahill, while you're in this file, there's a typo to fix: "Do not use the the"

File: docs/manage/shared/rolling-upgrades/_restart-impact.mdx [hunk 8]
```diff
@@ -0,0 +1,18 @@
+When brokers restart, clients may experience higher latency, nodes may experience CPU spikes when the broker becomes available again, and you may receives alerts about under-replicated partitions.
```
Comment (micheleRP) [inline docs/manage/shared/rolling-upgrades/_restart-impact.mdx:1]:
```suggestion
When brokers restart, clients may experience higher latency, nodes may experience CPU spikes when the broker becomes available again, and you may receive alerts about under-replicated partitions.
```

File: docs/manage/kubernetes/rolling-upgrade.mdx [hunk 9]
```diff
@@ -0,0 +1,494 @@
...
+  ...
+  ```
+
+  </details>
+
+  :::
+
+1. Check the [release notes](https://github.com/redpanda-data/redpanda/releases) to find information about what has changed between Redpanda versions.
+
+## Perform a Rolling Upgrade
```
Comment (micheleRP) [inline docs/manage/kubernetes/rolling-upgrade.mdx:145]:
```suggestion
## Perform a rolling upgrade
```

File: docs/manage/shared/rolling-upgrades/_restart-impact.mdx [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+1. Broker A shuts down.
+1. Client sends a request to broker A, and receives `NOT_LEADER_FOR_PARTITION`.
+1. Client requests metadata, and learns that the new leader is broker B.
+1. Client sends the request to broker B.
+
+### CPU spikes upon broker restart
+
+When a restarted broker becomes available again, you may see your nodes' CPU usage increase temporarily. This temporary increase in CPU usage is due to the cluster rebalancing the partition replicas.
+
+### Under-Replicated partitions
```
Comment (micheleRP) [inline docs/manage/shared/rolling-upgrades/_restart-impact.mdx:16]:
```suggestion
### Under-replicated partitions
```

**Review by micheleRP — APPROVED** [review]
lgtm

**Pull request merged**